rollout/trafficrouting/nginx/nginx.go

```diff
@@ -0,0 +1,230 @@
+		}
+	}
+
+	if len(desiredCanaryIngress.Spec.Rules) == 0 {
+		return nil, errors.New(fmt.Sprintf("ingress `%s` has no rules using service %s backend", stableIngressName, stableServiceName))
+	}
+
+	// Process additional annotations, would commonly be things like `canary-by-header` or `load-balance`
+	for k, v := range r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Nginx.AdditionalIngressAnnotations {
+		if !strings.HasPrefix(k, annotationPrefix) {
+			k = fmt.Sprintf("%s/%s", annotationPrefix, k)
+		}
+		desiredCanaryIngress.Annotations[k] = v
+	}
+	// Always set `canary` and `canary-weight` - `canary-by-header` and `canary-by-cookie`, if set,  will always take precedence
+	desiredCanaryIngress.Annotations[fmt.Sprintf("%s/canary", annotationPrefix)] = "true"
+	desiredCanaryIngress.Annotations[fmt.Sprintf("%s/canary-weight", annotationPrefix)] = fmt.Sprintf("%d", desiredWeight)
+
+	return desiredCanaryIngress, nil
+}
+
+// compareCanaryIngresses compares the current canaryIngress with the desired one and returns a patch
+func compareCanaryIngresses(current *extensionsv1beta1.Ingress, desired *extensionsv1beta1.Ingress) ([]byte, bool, error) {
+	// only compare Spec, Annotations, and Labels
+	return diff.CreateTwoWayMergePatch(
+		&extensionsv1beta1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: current.Annotations,
+				Labels:      current.Labels,
+			},
+			Spec: current.Spec,
+		},
+		&extensionsv1beta1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: desired.Annotations,
+				Labels:      desired.Labels,
+			},
+			Spec: desired.Spec,
+		}, extensionsv1beta1.Ingress{})
+}
+
+// Reconcile modifies Nginx Ingress resources to reach desired state
+func (r *Reconciler) Reconcile(desiredWeight int32) error {
+	stableIngressName := r.cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Nginx.StableIngress
+	canaryIngressName := ingressutil.GetCanaryIngressName(r.cfg.Rollout)
+
+	// Check if stable ingress exists (from lister, which has a cache), error if it does not
+	stableIngress, err := r.cfg.IngressLister.Ingresses(r.cfg.Rollout.Namespace).Get(stableIngressName)
+	if err != nil {
+		r.log.WithField(logutil.IngressKey, stableIngressName).WithField("err", err.Error()).Error("error retrieving stableIngress")
+		return errors.New(fmt.Sprintf("error retrieving stableIngress `%s` from cache: %v", stableIngressName, err))
+	}
+
+	// Check if canary ingress exists (from lister which has a cache), determines whether we later call Create() or Update()
+	canaryIngress, err := r.cfg.IngressLister.Ingresses(r.cfg.Rollout.Namespace).Get(canaryIngressName)
+
+	canaryIngressExists := true
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			// An error other than "not found" occurred
+			r.log.WithField(logutil.IngressKey, canaryIngressName).WithField("err", err.Error()).Error("error retrieving canary ingress")
+			return errors.New(fmt.Sprintf("error retrieving canary ingress `%s` from cache: %v", canaryIngressName, err))
```

Comment:
Feel free to ignore this, but what do you think about `fmt.Errorf` instead of `errors.New(fmt.Sprintf` to format errors?

Reply:
For some dumb reason I keep forgetting this exists - added.